### PR TITLE
*: support ACCOUNT (UN)LOCK when creating/altering user

### LINKED ISSUE #37051: Support ACCOUNT LOCK/UNLOCK for TiDB
# Motivation

In [MySQL](https://dev.mysql.com/doc/refman/8.0/en/account-locking.html), you can lock or unlock a user/role by `CREATE USER` or `ALTER USER`. For compatibility, TiDB should support this feature.

# Feature Acquirement

According to the description in [MySQL account locking](https://dev.mysql.com/doc/refman/8.0/en/account-locking.html), the feature is mainly related to `CREATE USER` and `ALTER USER` statements in TiDB.

* When used with [CREATE USER](https://docs.pingcap.com/tidb/stable/sql-statement-create-user#create-user), these clauses specify the initial locking state for a new account. In the absence of either clause, the account is created in an unlocked state.
* When used with [ALTER USER](https://docs.pingcap.com/tidb/stable/sql-statement-alter-user), these clauses specify the new locking state for an existing account. In the absence of either clause, the account locking state remains unchanged.
* Account locking state is recorded in the `account_locked` column of the `mysql.user` system table. The output from [SHOW CREATE USER](https://docs.pingcap.com/tidb/stable/sql-statement-show-create-user) indicates whether an account is locked or unlocked.
* If a client attempts to connect to a locked account, the attempt fails, and returns an `ErrAccountHasBeenLocked` error.
* The ability to use views is not affected by locking the account.
* After a role is created, the role is locked by default and cannot log in to TiDB; after unlocking through `ALTER USER`, the role can also log in normally.

Note that these behaviors are different from MySQL8:
* TiDB doesn't support the [`validate_password` component](https://dev.mysql.com/doc/mysql-security-excerpt/8.0/en/validate-password.html) (at least not in v6.1.0)
* TiDB doesn't support the [proxy user](https://dev.mysql.com/doc/refman/8.0/en/proxy-users.html) or [stored programs](https://dev.mysql.com/doc/dev/mysql-server/latest/stored_programs.html) (at least not in v6.1.0)
* TiDB would not support [*temporary lock*](https://dev.mysql.com/doc/mysql-security-excerpt/8.0/en/password-management.html#failed-login-tracking) when encountering too many consecutive login failures.

# Background

## Related Data Structures

For accelerating the reading of `mysql.user`, TiDB has a `type MySQLPrivilege struct` to cache the users' privilege:
```go
type MySQLPrivilege struct {
	User          []UserRecord
	UserMap       map[string][]UserRecord // Accelerate User searching
	Global        map[string][]globalPrivRecord
	Dynamic       map[string][]dynamicPrivRecord
	DB            []dbRecord
	DBMap         map[string][]dbRecord // Accelerate DB searching
	TablesPriv    []tablesPrivRecord
	TablesPrivMap map[string][]tablesPrivRecord // Accelerate TablesPriv searching
	ColumnsPriv   []columnsPrivRecord
	DefaultRoles  []defaultRoleRecord
	RoleGraph     map[string]roleGraphEdgesTable
}
```

The state of the account lock can be accessed by the `User` field, since `UserRecord` contains a field `AccountLocked bool`:
```go
type UserRecord struct {
	baseRecord

	AuthenticationString string
	Privileges           mysql.PrivilegeType
	AccountLocked        bool // A role record when this field is true
	AuthPlugin           string
}
```

## Update of Privilege

Each time to run a command related to change privilege, such as `CREATE USER`, `ALTER USER` and `FLUSH`, the function `NotifyUpdatePrivilege` would be called to:
1. update the local cache from `mysql.user`, and
2. notify the other instances to update by etcd

# Design

Since https://github.com/pingcap/tidb/pull/9377 has 
* imported `account_locked` column to the `mysql.user` table, which appeared first in an early version [v3.0.5](https://github.com/pingcap/tidb/releases/tag/3.0.5-hotfix), we don't have to modify the definition of `mysql.user` table. 
* imported `AccountLocked` field into `type UserRecord struct`, which is included in `MySQLPrivilege`, we don't have to modify thie definition of priviledge.

And the check of *lock* is placed in `func (p *UserPrivileges) ConnectionVerification`. The check of *lock* happens after the check of *password*.

# Testing

* The unit test is in the `func (cli *testServerClient) runTestAccountLock`
* The related mysql-test is also ported

### PR DESCRIPTION
**Pull request by CbcWestwolf**
### What problem does this PR solve?

Issue Number: close #37051  close https://github.com/pingcap/tidb/issues/35961

Problem Summary:

ACCOUNT (UN)LOCK has not been implemented.

### What is changed and how it works?

Lock/Unlock a user/role when creating/altering user.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed steps at last)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

#### Manual Test

1. start a cluster with 2 tidb-servers
2. using `root` to connect to tidb-server-1 and create a user `user1`, and `GRANT ALL on test.* to role1`
3. success to connect to tidb-server-2 using `user1`
4. run `alter user user1 account lock` in tidb-server-1
5. `user1` can not connect again to tidb-server-2 after it 'exit' the connection
6. run `alter user user1 account unlock` in tidb-server-1
7. `user1` can connect again to tidb-server-2

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Before, TiDB users cannot 'lock' a user by creating/altering user statements.

After this PR, TiDB users can do this like MySQL.
```

### CODE REVIEW

File: executor/simple.go [hunk 1]
```diff
@@ -793,8 +793,22 @@ func (e *SimpleExec) executeCreateUser(ctx context.Context, s *ast.CreateUserStm
...
+			} else if s.PasswordOrLockOptions[i].Type == ast.Unlock {
+				break
+			}
+		}
+	}
+
 	sql := new(strings.Builder)
-	if s.IsCreateRole {
+	if s.IsCreateRole || lockAccount {
 		sqlexec.MustFormatSQL(sql, `INSERT INTO %n.%n (Host, User, authentication_string, plugin, Account_locked) VALUES `, mysql.SystemDB, mysql.UserTable)
```
Comment (bb7133) [inline executor/simple.go:812]: It is kind of surprising to me that 'locked account' is used for RBAC. But anyway, I found MySQL 8 used it as the same way.

**Review by bb7133 — APPROVED** [review]
LGTM

**Comment by bb7133** [comment]
PTAL @tiancaiamao

File: executor/simple.go [hunk 2]
```diff
@@ -998,6 +1027,15 @@ func (e *SimpleExec) executeAlterUser(ctx context.Context, s *ast.AlterUserStmt)
 			}
 		}
 
+		if len(lockAccount) != 0 {
+			_, _, err := exec.ExecRestrictedSQL(ctx, nil,
+				`UPDATE %n.%n SET account_locked=%? WHERE Host=%? and User=%?;`,
+				mysql.SystemDB, mysql.UserTable, lockAccount, strings.ToLower(spec.User.Hostname), spec.User.Username)
```
Comment (morgo) [inline executor/simple.go:1033]: I think `spec.User.Hostname` should already be lowered in the parser? If not, no problem to do it here of course.

Comment (CbcWestwolf) [reply executor/simple.go]: Yes, the `ToLower` is unnecessary :-) FIxed it.

**Review by morgo — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]
Actually, you can split the PR for such kind.
A lot change is related to the `Auth()` interface which used to return boolean and now returns error...
The code change for ACCOUNT LOCK itself is not much.

**Pull request merged**